### ISSUE TITLE
New firmware support

### ISSUE THREAD

**Issue by marcosdvpereira** [intx82/a9-v720#16]
Hi.

My cam came with FW version: 202303131805.
Is there any way to support this version?

My setup:
* OpenWRT hijaking DNS requests to PiHole
* PiHole with DNS Record as recommend on your guide (PiHole dedicated to my IoT network only working with WhiteLists. Everything else is blocked)
* Docker Nginx container listening port 80 and 443. Added a domain based configuration as suggested [here](https://github.com/intx82/a9-v720/issues/15#issuecomment-1515783638)
* Docker Python container running fake server
* Docker MQTT container

No MQTT message is received nor receive feedback trying sending one.
Device list always empty.

Thanks any help here :)

**Comment by intx82** [comment]
Hello
Could you please make a capture of camera data? They may change hostnames of servers or entirely remove DNS resolve from camera. Check the domain names what camera try to resolve.

I saw they change an MQTT server host in version 20230301*, before was `p2p.v720.naxclow.com` after `v720.p2p.naxclow.com`

**Comment by marcosdvpereira** [comment]
I've followed all steps on documentation, and as soon I connected it to IoT SSID it's only try to resolve `v720.naxclow.com` once and nothing more happens. The light of "mode" button blinks until no battery remains, and that's it!
I successfully can connect directly to the cam using the python script, but no luck using routers WiFi.

My goal here is trying to use the cam on Home Assistant without wan/cloud access.

Is there any useful info that I can take from the cam?

**Comment by intx82** [comment]
i would be appreciated if you attach somewhere PCAP file and verbose logs from server, if content is not secret of course. 

Did you seen HTTP request to the server from camera? (you may add '-v' argument to the server to retrieve more logs)

**Comment by grrosminet** [comment]
Hi

Same firmware here

the camera connects to wifi, then I got two DNS requests and the mode light becomes still
```
May  2 11:54:25 pi4 dnsmasq[22625]: query[A] v720.naxclow.com from 192.168.5.35
May  2 11:54:25 pi4 dnsmasq[22625]: query[A] v720.p2p.naxclow.com from 192.168.5.35
```

In MQTT broker I got two topics : 
1) Naxclow/P2P/Users/Device/Status  :  {"device":"0800c00A4DF4","token":"NAXCLOW","status": 1}
2) Naxclow/P2P/Users/Device/Info : {"apStatus": 0,"devPower": 69,"IrLed": 1,"sdCapacity": 0,"SD_freeDisk": -1,"SD_blockDisk": -1,"sdMoveMode": 0,"sdCardReco": 1,"sd_State": 0,"cameraState": 0,"instLed": 1,"random":"UVWXYZ","token":"ca9c11a604","devicesCode":"0800c00A4DF4","wifiName":"Domotique","version":"202303131805"}

Logs from the fake server are empty ... 
```
$ cat nohup.out 
2023-05-02 10:17:01,700  [   INFO] [TCP-SRV :6123] Waiting for connection
2023-05-02 10:17:01,721  [   INFO] [UDP-SRV :6123] Waiting for connection
2023-05-02 10:24:37,860  [   INFO] [HTTP] GET device list: /dev/list
2023-05-02 10:24:37,862  [   INFO] [HTTP] "GET /dev/list HTTP/1.1" 200 -

```

devices list is empty ...
```
$ wget -q http://127.0.0.1/dev/list -O -
[]
```

the only request from camera trying to get out of network is the DNS request (iptables logs)

what can I provide you to help ?

**Comment by dbuezas** [comment]
I also have the same fw version. I saw the camera tries to resolve these dns names:
* caps.xmcsrv.net
* pub-cfg.secu100.net

**Comment by dbuezas** [comment]
It just started working by accident!
Since adguard was reporting all dns requests as coming from the same client (my router's IP), I searched and found that I should _Set your adguard instance as the DNS server that is advertised over DHCP, not as the global DNS server of the router_ [in reddit](https://www.reddit.com/r/Adguard/comments/nozvjf/home_fritzbox_resolving_local_domain_names_clients/).
I don't understand how that would make a difference, and maybe it was just a matter of waiting, but I can now see the video stream.
Here I see the fw version to be the one reported in this issue:
<img width="516" alt="image" src="https://user-images.githubusercontent.com/777196/236700958-a51ad92c-9f61-4035-b5a5-05c51508db88.png">
And here the output of the server now:
<img width="563" alt="image" src="https://user-images.githubusercontent.com/777196/236701046-2fef4337-dd1e-4516-8468-b3409ae194a3.png">

My setup:
* AdGuard doing the DNS rewrites (as a HomeAssistant AddOn)
* Fritz!Box Router
* My Mac running this "fake server"

**Comment by dbuezas** [comment]
I found something interesting: I had p2p.v720.naxclow.com redirected to broker.hivemq.com, but had also cut the a9 camera from the internet. This means that it couldn't connect to the mqtt server at all.
If I give it internet access, the device disconnects:
<img width="855" alt="image" src="https://github.com/intx82/a9-v720/assets/777196/caaa028d-1133-4716-9b30-d0113564763f">
Interestingly, the light stays on, so it thinks it is still connected to the fake server. If I kill the fake server, it starts blinking again.

It looks like something happens in the new firmware when the device can reach out to to the mqtt broker

**Comment by dbuezas** [comment]
Ok, that may have been only a coincidence. Now it is very unstable regardless of what I do

**Comment by dbuezas** [comment]
Ok I found something more, Internet access may be irrelevant but this seems to hold:
If the camera is turned off and on again, it connects again but the error above appears when I try to get a stream or snapshot. To make it work again I need to kill the fake-server, turn the camera off, start the fake-server again and then the camera.

**Comment by dbuezas** [comment]
I got another camera which happens to have the older firmware today so I have been experimenting a bit more:
* The camera with the old firmware reconnects quickly and it is generally faster to recover.
* The new camera eventually reconnects, it just takes a while (after either a cam restart or a server restart), but it eventually connects.

This means that at least for me, this software works on both versions, it is just better on the old firmware. On the bright side, the new FW has much better picture quality and wider FOV.

I have been playing around with making a Home Assistant addon and doing some ffmpeg reencoding, and I ended up with a slightly contrived but reliably working system, I'll post the AddOn soon.

The only thing I'd like to add is the ability to get status and send commands from/to the camera. I know this can be done through mqtt (and it works really well :) ), the issue is that the local mqtt broker one would have in HA is not an anonymous one. This means one needs to either use a public one in the internet or just live without it.

Seeing that commands are sent during the initial handshake, I hacked myself an extra endpoint in the http server meant to get the list of devices, snapshot and live stream. This one sends this json to the camera via both udp and tcp open connections, but the camera doesn't switch to IR mode. Do you know if this is possible without MQTT?
I added this to the `v720_sta.py` file:
```
def fsm_send_any(self, json_string: str):
        import json as jsonlib
        resp = prot_json_udp(json=jsonlib.loads(json_string))
        print(f'[TCP] Send any: {resp}')
        self._tcp.send(resp.req())
        print(f'[UDP] Send any: {resp}')
        self._udp.send(resp.req())
```
And call it from `v720_http.py`.

I get this in the console
<img width="142" alt="image" src="https://github.com/intx82/a9-v720/assets/777196/95a087d4-e2f1-40fd-8d2d-137a4e78ca3c">
but the video feed doesn't switch to black and white. Any tips?

**Comment by dbuezas** [comment]
Ok, I found how to send them via TCP.
BTW, it looks like the new fw version does work, it just takes a very long time to go online. Once that happens (and as long as the server is still running) it is quite stable.
Also the video and audio quality is much better.

**Comment by intx82** [comment]
Sorry guys, I have a vacation and was off. :)

Nice to hear that you have solved yours issues, if you have any changes into code, please make pull-request

About MQTT topics, check file `fake-server.md` there were examples how to put it back into color mode or other way put it into AP mode and use argument `--ir` (ie send this command from AP mode, not from MQTT)

**Comment by dbuezas** [comment]
After researching your code for a while, I found out how to send commands via UDP and the IR mode can be sent via STA mode. I'll eventually add entities for this to the Home Assistant Add-on I just made :) 
https://github.com/dbuezas/a9-camera-ha-add-on
Still very rough, but it is functional.

I managed to get the audio stream working in sta mode and merge it with the video stream using ffmpeg. Here's the relevant code:
https://github.com/dbuezas/a9-camera-ha-add-on/blob/main/rootfs/usr/local/a9-v720/src/v720_http.py#L73-L166

I had to butcher your code to remove the dependencies that don't work well with the alpine docker base, so I just put it in that repo. If you want I can make a PR with the new stream api (it works fine in vlc too, with audio and all :) )

Thanks for this repo! Considering the effort it took me to just do that, I have gained a lot of respect for the work you did here reverse engineering everything bout this cam 🙇
